r/R/dplyr.R

```diff
@@ -279,6 +280,19 @@ arrow_eval <- function (expr, mask) {
     }
     invisible(structure(msg, class = "try-error", condition = e))
   })
+  # Check that the evaluated expression has one of the expected classes or types
+  if (!inherits(out, c("array_expression", "Expression", "try-error")) &&
+      !typeof(out) %in% c("logical", "integer", "double", "character", "NULL")) {
```

Comment:
I think this is too restrictive--what about "list", "raw", etc., which can convert to Arrow objects? What if the input is an `Array`?

Another approach might be to use a blocklist and only block certain cases (like the `function` you ran into).

I'm also not confident that `arrow_eval` is the right place for this. It very well might be, but maybe there are applications of `arrow_eval` that would, for example, return a function or an environment.

Reply:
I'll dig into this more to be sure. Whether we use a passlist or blocklist for the types is probably not too important because there is a [finite and relatively small number of possible types](https://github.com/wch/r-source/blob/trunk/src/main/util.c#L192-L222).

FYI check out [`rlang::is_syntactic_literal()`](https://rlang.r-lib.org/reference/is_expression.html) which is defined like so:
```r
function (x) 
{
    switch(typeof(x), `NULL` = {
        TRUE
    }, logical = , integer = , double = , character = {
        length(x) == 1
    }, complex = {
        if (length(x) != 1) {
            return(FALSE)
        }
        is_na(x) || Re(x) == 0
    }, FALSE)
}
```
That was my starting point for this. But I agree we should at least add `"raw"` to that.

Reply:
I hadn't realized you can do this
```
y <- Array$create(c(1,2,3))
tibble::tibble(x = c("a","b","c")) %>% Table$create() %>% mutate(y)
```
Very cool. So yes the class list would need to be expanded.

I'll check again, but I don't think we have any tests that exercise that, because adding the checks above didn't cause any tests to fail.

Reply:
I'll move these checks into a separate function and call it after `arrow_eval()` in the relevant dplyr verb methods.